stdlib/atomic.mli

```diff
@@ -17,6 +17,84 @@
 
 (** Atomic references.
 
+    A basic use case is to have global counters that are updated in a
+    thread-safe way:
+
+    {[
+    (* our counter *)
+    let count_bytes_read = Atomic.make 0;;
+
+    (* prepare a sample file *)
+    let () =
+      let oc = open_out "/tmp/example_data" in
+      for i=1 to 100_000 do output_char oc 'x' done;
+      close_out oc ;;
+
+    (* just read from file, discard content but count bytes.
+       This is pretty useless and only used to show this. *)
+    let read_file () =
+      let ic = open_in "/tmp/example_data" in
+      let buf = Bytes.create 1024 in
+
+      let rec loop () =
+        let n = input ic buf 0 1024 in
+        Thread.yield();
+        if n> 0 then (
+          (* count_bytes_read += n, atomically *)
+          ignore (Atomic.fetch_and_add count_bytes_read n : int);
+          loop()
+        )
+      in
+      loop()
+    ;;
+
+    (* run multiple threads (or domains) that update the counter *)
+    # let () =
+      let threads = Array.init 8 (fun _ -> Thread.create read_file ()) in
+      Array.iter Thread.join threads;
+      Printf.printf "read %d bytes\n" (Atomic.get count_bytes_read)
```

Comment:
I find it a bit awkward to exemplify concurrent atomics with a single-domain example that does not actually need atomics. Alternatives include using `Domain` directly (con: arguably too low-level in general) or domainslib (con: outside the stdlib, may change or get replaced in the medium-term future).

Reply:
I've been testing it in a 4.14 repl on the side (in which it was surprisingly hard to get the `Thread` toplevel module?!), so I don't know exactly how I should go about that. Maybe I just need a 5.0-alpha switch to run this.

`Domain.spawn` should be on par in terms of complexity with `Thread.create`, right? Ignoring the ticker thread…

Reply:
Another option is to show not a complete program with Domain.spawn, but just the code of a library function that *may* be called concurrently. (For example: a "unique id" generator that works by incrementing an internal counter.) You can show the sequential version, point out that it is not domain-safe, and move to Atomic instead.

Reply:
removed

Reply:
moved to `Domain`. Hopefully someone can check the code, unless I get to trying 5.0-alpha on that.